rsyslog/undefined_field/undefined_field.go

```diff
@@ -0,0 +1,392 @@
+			}
+			if val, ok := getMapStringValue(configMap, "CDM_UNDEFINED_MAX_NUM_FIELDS"); ok {
+				var rval64 int64
+				if rval64, err = strconv.ParseInt(val, 10, 64); err == nil {
+					undefined_max_num_fields = rval64
+				} else {
+					fmt.Fprintln(logfile, "ERROR: Invalid CDM_UNDEFINED_MAX_NUM_FIELDS value [%s]: [%v]", val, err)
+					panic(fmt.Errorf("Invalid CDM_UNDEFINED_MAX_NUM_FIELDS value [%s]: [%v]", val, err))
+				}
+			}
+		}
+		defer config.Close()
+	} else if !strings.Contains(err.Error(), "no such file or directory") {
+		fmt.Fprintln(logfile, "ERROR: Could not open config file [%s]: [%v]", undefined_config, err)
+		panic(fmt.Errorf("Could not open config file [%s]: [%v]", undefined_config, err))
+	}
+
+	if !use_undefined && tmp_keep_empty_fields == "" && !undefined_to_string && undefined_dot_replace_char == "UNUSED" {
+		noaction = true
+	}
+	if undefined_max_num_fields == -1 {
+		undefined_max_num_fields = int64(^uint(0) >> 1)
+	}
+	if use_undefined {
+		tmp_default := strings.Split(default_keep_fields, ",")
+		tmp_extra := strings.Split(extra_keep_fields, ",")
+		keep_fields = make(map[string]string)
+		for _, str := range tmp_default {
+			keep_fields[str] = str
+		}
+		for _, str := range tmp_extra {
+			keep_fields[str] = str
+		}
+	}
+	tmp := strings.Split(tmp_keep_empty_fields, ",")
+	keep_empty_fields = make(map[string]string)
+	for _, str := range tmp {
+		keep_empty_fields[str] = str
+	}
+
+	if undefined_dot_replace_char != "UNUSED" {
+		replacer = strings.NewReplacer(".", undefined_dot_replace_char)
+	}
+
+	fmt.Fprintln(logfile, "mmexternal: merge_json_log: ", merge_json_log)
+	fmt.Fprintln(logfile, "mmexternal: use_undefined: ", use_undefined)
+	fmt.Fprintln(logfile, "mmexternal: default_keep_fields: ", default_keep_fields)
+	fmt.Fprintln(logfile, "mmexternal: extra_keep_fields: ", extra_keep_fields)
+	fmt.Fprintln(logfile, "mmexternal: undefined_name: ", undefined_name)
+	fmt.Fprintln(logfile, "mmexternal: keep_empty_fields: ", tmp_keep_empty_fields)
+	fmt.Fprintln(logfile, "mmexternal: undefined_to_string: ", undefined_to_string)
+	fmt.Fprintln(logfile, "mmexternal: undefined_dot_replace_char: ", undefined_dot_replace_char)
+	fmt.Fprintln(logfile, "mmexternal: undefined_max_num_fields: ", undefined_max_num_fields)
+	fmt.Fprintln(logfile, "mmexternal: noaction: ", noaction)
+}
+
+func replaceDotMoveUndefined(jsonMap map[string]interface{}, topPropLevel bool) (map[string]interface{},bool,bool) {
+	replace_me := false
+	has_undefined := false
+	cp := make(map[string]interface{})
+	for origkey, value := range jsonMap {
+		key := origkey
+		if topPropLevel && merge_json_log && undefined_dot_replace_char != "UNUSED" {
+			// replace '.' with specified char (e.g., '_')
+			key = replacer.Replace(origkey)
+			if key != origkey {
+				replace_me = true
+			}
+		}
+		// skip empty or not?
+		valuemap, ismap := value.(map[string]interface{})
+		valuearraymap, isarraymap := value.([]map[string]interface{})
```

Comment:
I think this should be `value.([]interface{})`

Reply:
Ok...